code/game/machinery/bots/cprbot.dm

```diff
@@ -0,0 +1,332 @@
+		human = WEAKREF(patient)
+		if (patient && !(patient in has_said_to_patient))
+			visible_message("[patient] is injured! I'm coming!")
+			has_said_to_patient += patient
+
+		move_to_target()
+	else
+		// If no patients are found, check if owner is nearby and follow them if idle
+		if (state == STATE_CPRBOT_IDLE && (owner && (owner in view(search_radius, src))))
+			state = STATE_CPRBOT_FOLLOWING_OWNER
+			walk_to(src, owner, 0, movement_delay)
+		else if (state == STATE_CPRBOT_FOLLOWING_OWNER)
+			// Continue following the owner if no patient is in sight
+			walk_to(src, owner, 0, movement_delay)
+		else
+			go_idle()
+
+/obj/structure/machinery/bot/cprbot/proc/can_still_see_patient()
+	if (human == null)
+		return FALSE
+
+	var/mob/living/carbon/human/patient = human.resolve()
+	if (patient == null)
+		return FALSE
+
+	return patient in view(search_radius, src)
+
+/obj/structure/machinery/bot/cprbot/proc/patient_in_range()
+	if (human == null)
+		return FALSE
+
+	var/mob/living/carbon/human/patient = human.resolve()
+	if (patient == null)
+		return FALSE
+
+	return get_dist(src, patient) == 0
+
+/obj/structure/machinery/bot/cprbot/proc/move_to_target()
+	var/mob/living/carbon/human/patient = human.resolve()
+	// If we cannot see them anymore then stop moving
+	if (!can_still_see_patient())
+		go_idle()
+		return
+
+	// It might not exist anymore or something
+	if (patient == null)
+		go_idle()
+		return
+
+	if (!patient_in_range())
+		// We are already moving
+		if (state == STATE_CPRBOT_MOVING)
+			return
+
+		state = STATE_CPRBOT_MOVING
+		walk_to(src, patient, 0, movement_delay)
+	else
+		walk_to(src, 0) // make sure we stop walking
+		state = STATE_CPRBOT_CPR
+		switch_to_faster_processing()
+		try_perform_cpr()
+
+/obj/structure/machinery/bot/cprbot/proc/perform_cpr(mob/living/carbon/human/target)
+	if (!cpr_ready)
+		return
+
+	currently_healing = TRUE
+
+	update_icon()
+
+	target.revive_grace_period += 4 SECONDS
+	target.visible_message(SPAN_NOTICE("<b>[src]</b> automatically performs <b>CPR</b> on <b>[target]</b>."))
+	target.balloon_alert_to_viewers("Performing CPR, stay clear!")
+	currently_healing = TRUE
+	playsound(loc, 'sound/CPRbot/CPR.ogg', 25, 1)
+	cpr_ready = FALSE
+	addtimer(VARSET_CALLBACK(src, cpr_ready, TRUE), 7 SECONDS)
+
+/obj/structure/machinery/bot/cprbot/proc/try_perform_cpr()
+	currently_healing = TRUE
+	// Resolve the weak reference to check if the target still exists
+	var/mob/living/carbon/human/target = human.resolve()
+
+	if (!patient_in_range())
+		go_idle()
+		switch_to_slower_processing()
+		return
+
+	// Check if the target is valid and still needs CPR
+	if ((target.stat != DEAD) || (target.stat == DEAD && !target.check_tod()))
+		go_idle()
+		switch_to_slower_processing()
+		return
+
+	perform_cpr(target)
+
+/obj/structure/machinery/bot/cprbot/proc/self_destruct(mob/living/carbon/human/user = null)
+	var/obj/item/cprbot_item/I = new /obj/item/cprbot_item(src.loc)
```

Comment:
should replace `I` with a more descriptive name here (reusing subtype `cprbot_item` would be totally fine imo, i.e. `var/obj/item/cprbot_item/cprbot_item`)